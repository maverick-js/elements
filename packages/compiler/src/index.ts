export * from './plugins';
export * from './transformer';
