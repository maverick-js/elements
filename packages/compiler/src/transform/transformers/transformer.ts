import type ts from 'typescript';

import type { AstNode } from '../../parse/ast';
import type { TransformContext } from '../transform';

export interface Transform {
  (data: TransformData): ts.SourceFile;
}

export interface TransformData {
  sourceFile: ts.SourceFile;
  nodes: AstNode[];
  ctx: TransformContext;
}

export interface StateTransform<State> {
  (node: AstNode, state: State): StateTransformResult;
}

export type StateTransformResult = ts.Expression | Array<ts.Expression | ts.Statement> | undefined;
