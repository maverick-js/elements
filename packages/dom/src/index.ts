export * from './insert';
export * from './reconcile';
export * from './render';
export * from './runtime';
export * from './walker';

export * from './components/for';
export * from './components/fragment';
export * from './components/host';
export * from './components/portal';
