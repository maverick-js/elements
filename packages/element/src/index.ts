export * from './create';
export * from './element';
export * from './types';
export * from './define';
export * from './symbols';
