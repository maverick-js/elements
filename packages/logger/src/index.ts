export * from './logger';
export * from './ms';
