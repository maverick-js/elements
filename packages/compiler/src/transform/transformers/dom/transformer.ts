import {
  $,
  createImports,
  isJsxRootNode,
  resetUniqueIdCount,
  splitImportsAndBody,
  transformTsNode,
} from '@maverick-js/ts';
import ts from 'typescript';

import {
  isMaverickCoreImportDeclaration,
  removeVirtualComponentImports,
} from '../../../parse/analysis';
import { Scope } from '../../../parse/ast';
import { parse } from '../../../parse/parse';
import { setupCustomElements } from '../shared/element';
import { Variables } from '../shared/variables';
import type { Transform, TransformData } from '../transformer';
import type { DomRuntime } from './runtime';
import { DomTransformState } from './state';
import { transform } from './transform';

export interface DomTransformOptions {
  /**
   * Whether to mark components with a static `element` property declaration to be
   * built as a custom element.
   */
  customElements?: boolean;
  /**
   * When set to `true`, the DOM runtime will connect the server-rendered HTML with the JS on
   * the client, making it interactive.
   *
   * The runtime does not require the component to be rendered on the server first but it's
   * recommended for the best performance.
   */
  hydratable?: boolean;
  /**
   * A single event listener is attached to the document to manage expensive events on multiple
   * child elements. Instead of adding individual listeners to each child, the event "bubbles" up
   * from the target element to the document, where the listener can capture and handle it. This
   * improves performance, especially in dynamic interfaces, by reducing the number of event
   * listeners.
   */
  delegateEvents?: boolean;
}

export function createDomTransform(options?: DomTransformOptions): Transform {
  return (data) => domTransform(data, options);
}

export function domTransform(
  { sourceFile }: TransformData,
  { customElements, hydratable, delegateEvents }: DomTransformOptions = {},
) {
  let state = new DomTransformState(null, { hydratable }),
    seenImports = false;

  function visit(this: ts.TransformationContext, node: ts.Node) {
    if (!seenImports && isMaverickCoreImportDeclaration(node)) {
      seenImports = true;
      return removeVirtualComponentImports(node, (name) => state.runtime.add(name));
    } else if (isJsxRootNode(node)) {
      const ast = parse(node),
        result = ast ? transform(ast, state.child(ast, new Scope())) : $.null;

      resetUniqueIdCount();

      return result;
    }

    return ts.visitEachChild(node, visit, this);
  }

  const transformedSourceFile = transformTsNode(sourceFile, (ctx) => visit.bind(ctx)),
    { imports, body } = splitImportsAndBody(transformedSourceFile);

  if (delegateEvents && state.delegatedEvents.size > 0) {
    body.push(createDelegateEventsStatement(state.delegatedEvents, state.runtime));
  }

  const statements: ts.Statement[] = [];

  const templates = createTemplateVariables(state);

  if (state.runtime.identifiers.length > 0) {
    statements.push(state.runtime.toImportDeclaration());
  }

  if (templates) statements.push(templates.toStatement());

  if (state.vars.module.length > 0) {
    statements.push(state.vars.module.toStatement());
  }

  const updatedSourceFile = $.updateSourceFile(transformedSourceFile, [
    ...imports,
    ...statements,
    ...state.renders,
    ...body,
  ]);

  return customElements
    ? setupCustomElements(updatedSourceFile, registerCustomElement, [createElementImport()])
    : updatedSourceFile;
}

const registerCustomElementExpression = $.call($.id('$$_define_custom_element'), [$.id('this')]);

function registerCustomElement() {
  return registerCustomElementExpression;
}

function createElementImport() {
  return createImports([$.id('$$_define_custom_element')], '@maverick-js/element');
}

function createTemplateVariables(state: DomTransformState) {
  const templates: [template: ts.Identifier, html: string, importNodes: boolean][] = [],
    vars = new Variables();

  state.walk(({ template, html, importNodes }) => {
    if (template && html) templates.push([template, html, importNodes]);
  });

  for (const [template, html, importNodes] of templates) {
    vars.create(template, state.runtime.createTemplate($.string(html), importNodes));
  }

  // Dedupe templates.
  for (let i = 0; i < templates.length; i++) {
    const templateA = templates[i];
    for (let j = i + 1; j < templates.length; j++) {
      const templateB = templates[j];
      if (templateA[1] === templateB[1] && templateA[2] === templateB[2]) {
        vars.update(templateB[0], templateA[0]);
      }
    }
  }

  return templates.length > 0 ? vars : null;
}

export function createDelegateEventsStatement(events: Set<string>, runtime: DomRuntime) {
  return $.createExpressionStatement(
    runtime.delegateEvents($.array(Array.from(events).map((type) => $.string(type)))),
  );
}
