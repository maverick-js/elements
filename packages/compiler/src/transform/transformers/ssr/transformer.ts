import { isArray } from '@maverick-js/std';
import {
  $,
  addStatementsToBlock,
  type BlockUpdate,
  createStatements,
  isJsxRootNode,
  replaceTsNodes,
  splitImportsAndBody,
  transformTsNode,
  updateBlock,
} from '@maverick-js/ts';
import ts from 'typescript';

import {
  isMaverickCoreImportDeclaration,
  removeVirtualComponentImports,
} from '../../../parse/analysis';
import { parse } from '../../../parse/parse';
import { setupCustomElements } from '../shared/element';
import { returnLastExpression } from '../shared/factory';
import type { Transform, TransformData } from '../transformer';
import { SsrTransformState } from './state';
import { transform } from './transform';

export interface SsrTransformOptions {
  /**
   * Whether to mark components with a static `element` property declaration to be
   * rendered as a custom element.
   */
  customElements?: boolean;
}

export function createSsrTransform(options?: SsrTransformOptions): Transform {
  return (data) => ssrTransform(data, options);
}

export function ssrTransform(
  { sourceFile }: TransformData,
  { customElements }: SsrTransformOptions = {},
) {
  let state = new SsrTransformState(null),
    seenImports = false,
    blocks: BlockUpdate[][] = [];

  function visit(this: ts.TransformationContext, node: ts.Node) {
    if (!seenImports && isMaverickCoreImportDeclaration(node)) {
      seenImports = true;
      return removeVirtualComponentImports(node, (name) => state.runtime.add(name));
    } else if (ts.isBlock(node) || ts.isModuleBlock(node)) {
      const updates: BlockUpdate[] = [];

      blocks.push(updates);
      const block = ts.visitEachChild(node, visit, this);
      blocks.pop();

      return updateBlock(block, updates);
    } else if (isJsxRootNode(node)) {
      const ast = parse(node),
        result = ast ? transform(ast, state.child(ast)) : $.null,
        currentBlock = blocks.at(-1);

      if (currentBlock) {
        if (isArray(result)) {
          const setup = result.slice(0, -1);

          currentBlock.push({
            type: 'add',
            statements: setup,
            before: node,
          });

          currentBlock.push({
            type: 'replace',
            old: node,
            new: result.at(-1)!,
          });
        } else {
          currentBlock.push({
            type: 'replace',
            old: node,
            new: result,
          });
        }

        // Return original block and transform above.
        return node;
      }

      if (isArray(result)) {
        returnLastExpression(result);
        return $.selfInvokedFn(result);
      }

      return result;
    }

    return ts.visitEachChild(node, visit, this);
  }

  const transformedSourceFile = transformTsNode(sourceFile, (ctx) => visit.bind(ctx)),
    { imports, body } = splitImportsAndBody(transformedSourceFile);

  const statements: ts.Statement[] = [];

  if (state.runtime.identifiers.length > 0) {
    statements.push(state.runtime.toImportDeclaration());
  }

  addTemplateVariables(state);
  if (state.vars.module.length > 0) {
    statements.push(state.vars.module.toStatement());
  }

  const updatedSourceFile = $.updateSourceFile(transformedSourceFile, [
    ...imports,
    ...statements,
    ...body,
  ]);

  return customElements
    ? setupCustomElements(updatedSourceFile, registerCustomElement)
    : updatedSourceFile;
}

function registerCustomElement() {
  return $.createTrue();
}

function addTemplateVariables(state: SsrTransformState) {
  state.walk(({ template, statics }) => {
    if (template && statics.length > 0) {
      state.vars.module.create(template, $.array(statics));
    }
  });
}
