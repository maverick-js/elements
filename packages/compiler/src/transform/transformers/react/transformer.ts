import { isArray } from '@maverick-js/std';
import {
  $,
  addStatementsToBlock,
  type BlockUpdate,
  isJsxRootNode,
  replaceTsNode,
  resetUniqueIdCount,
  splitImportsAndBody,
  transformTsNode,
  updateBlock,
} from '@maverick-js/ts';
import ts from 'typescript';

import {
  isMaverickCoreImportDeclaration,
  removeVirtualComponentImports,
} from '../../../parse/analysis';
import { parse } from '../../../parse/parse';
import { createDelegateEventsStatement } from '../dom/transformer';
import { returnLastExpression } from '../shared/factory';
import type { Transform, TransformData } from '../transformer';
import { ReactTransformState } from './state';
import { transform } from './transform';

export interface ReactTransformOptions {
  /**
   * A single event listener is attached to the document to manage expensive events on multiple
   * child elements. Instead of adding individual listeners to each child, the event "bubbles" up
   * from the target element to the document, where the listener can capture and handle it. This
   * improves performance, especially in dynamic interfaces, by reducing the number of event
   * listeners.
   */
  delegateEvents?: boolean;
}

export function createReactTransform(options?: ReactTransformOptions): Transform {
  return (data) => reactTransform(data, options);
}

export function reactTransform(
  { sourceFile }: TransformData,
  { delegateEvents }: ReactTransformOptions = {},
) {
  let state = new ReactTransformState(null),
    seenImports = false,
    blocks: BlockUpdate[][] = [];

  function visit(this: ts.TransformationContext, node: ts.Node) {
    if (!seenImports && isMaverickCoreImportDeclaration(node)) {
      seenImports = true;
      return removeVirtualComponentImports(node, (name) => state.runtime.add(name));
    } else if (ts.isBlock(node) || ts.isModuleBlock(node)) {
      const updates: BlockUpdate[] = [];

      blocks.push(updates);
      const block = ts.visitEachChild(node, visit, this);
      blocks.pop();

      return updateBlock(block, updates);
    } else if (isJsxRootNode(node)) {
      const ast = parse(node),
        result = ast ? transform(ast, state.child(ast)) : $.null,
        currentBlock = blocks.at(-1);

      resetUniqueIdCount();

      if (currentBlock) {
        if (isArray(result)) {
          const statements = result.slice(0, -1);

          currentBlock.push({
            type: 'add',
            statements,
            before: node,
          });

          currentBlock.push({
            type: 'replace',
            old: node,
            new: result.at(-1)!,
          });
        } else {
          currentBlock.push({
            type: 'replace',
            old: node,
            new: result,
          });
        }

        // Return original block and transform above.
        return node;
      }

      if (isArray(result)) {
        returnLastExpression(result);
        return $.selfInvokedFn(result);
      }

      return result;
    }

    return ts.visitEachChild(node, visit, this);
  }

  const transformedSourceFile = transformTsNode(sourceFile, (ctx) => visit.bind(ctx)),
    { imports, body } = splitImportsAndBody(transformedSourceFile);

  if (delegateEvents && state.delegatedEvents.size > 0) {
    body.push(createDelegateEventsStatement(state.delegatedEvents, state.domRuntime));
  }

  const statements: ts.Statement[] = [];

  const runtimes = [state.runtime, state.domRuntime, state.ssrRuntime];
  for (const runtime of runtimes) {
    if (runtime.identifiers.length > 0) {
      statements.push(runtime.toImportDeclaration());
    }
  }

  if (state.module.vars.length > 0) {
    statements.push(state.module.vars.toStatement());
  }

  return $.updateSourceFile(sourceFile, [
    ...imports,
    ...statements,
    ...state.module.block,
    ...body,
  ]);
}
