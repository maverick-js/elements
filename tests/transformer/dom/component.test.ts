import { transform } from 'src/transformer';

const t = (code: string) => transform(code).code;

it('should compile root expression ', () => {
  const result = t(`
function Component() {
  return id > 10 ? <div>{id}</div> : 20;
}
  `);
  expect(result).toMatchSnapshot();
});

it('should compile root component', () => {
  const result = t(`<Component />`);
  expect(result).toMatchSnapshot();
});

it('should compile child component', () => {
  const result = t(`<div><Component /></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile child component with props', () => {
  const result = t(`<div><Component foo="a" bar={id()} /></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile child component with spread', () => {
  const result = t(`<div><Component {...props} /></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile child component with multiple spreads', () => {
  const result = t(`<div><Component {...props} {...propsTwo} {...propsThree} /></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile child component with props and spread', () => {
  const result = t(`<div><Component foo="..." bar={id() + 10} {...props} /></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile child component with text children', () => {
  const result = t('<Component>foo 10 bar 20 baz</Component>');
  expect(result).toMatchSnapshot();
});

it('should compile child component with element children', () => {
  const result = t(`<div><Component><div></div></Component></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile child component with props and children', () => {
  const result = t(`<div><Component foo={id}><div></div></Component></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile child component containing expression', () => {
  const result = t(`<div><Component>{() => <div>{id()}</div>}</Component></div>`);
  expect(result).toMatchSnapshot();
});

it('should forward single call expression', () => {
  const result = t(`<div><Component>{() => <div></div>}</Component></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile child component with fragment children', () => {
  const result = t(
    `<div><Component><><div id="foo"></div><div id="bar"></div></></Component></div>`,
  );
  expect(result).toMatchSnapshot();
});

it('should insert multiple child components', () => {
  const result = t(`
<div>
  <Component />
  <Component />
  <div>Foo</div>
  <Component />
</div>`);
  expect(result).toMatchSnapshot();
});

it('should insert multiple fragmented child components', () => {
  const result = t(`
<div>
  <>
    <Component />
    <Component />
    <div>Foo</div>
    <Component />
  </>
</div>`);
  expect(result).toMatchSnapshot();
});
