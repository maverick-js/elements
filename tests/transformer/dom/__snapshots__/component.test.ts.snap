// Vitest Snapshot v1

exports[`should compile child component 1`] = `
"import { $$_clone, $$_insert, $$_create_component, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_insert($$_el, $$_create_component(Component));

  return $$_el;
})()"
`;

exports[`should compile component child fragment 1`] = `
"import { $$_clone, $$_insert, $$_create_template, $$_create_component } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div id=\\"foo\\"></div><div id=\\"bar\\"></div><div></div>\`);

$$_create_component(Component, {
  get children() {
    const $$_root = $$_clone($$_templ),
      $$_el = $$_root.firstChild,
      $$_el_2 = $$_el.nextSibling,
      $$_el_3 = $$_el_2.nextSibling;

    $$_insert($$_el, id);
    $$_insert($$_el_2, id);
    $$_insert($$_root, id, $$_el_3);
    $$_insert($$_root, id, null);

    return $$_root;
  },
})
"
`;

exports[`should compile component with element children 1`] = `
"import { $$_clone, $$_insert, $$_create_template, $$_create_component } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div>Foo</div>\`);
$$_create_component(Component, {
  get children() {
    const $$_root = $$_clone($$_templ),
      $$_el = $$_root.firstChild;

    $$_insert($$_el, id, null);

    return $$_el;
  },
})"
`;

exports[`should compile component with multiple spreads 1`] = `
"import { $$_create_component, $$_merge_props } from \\"maverick.js/dom\\";
$$_create_component(Component, $$_merge_props(props, propsTwo, propsThree))"
`;

exports[`should compile component with props 1`] = `
"import { $$_create_component } from \\"maverick.js/dom\\";
$$_create_component(Component, { foo: \\"a\\", bar: 10, boo: true, baz: id, qux: true })"
`;

exports[`should compile component with props and children 1`] = `
"import { $$_create_template, $$_clone, $$_create_component } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
$$_create_component(Component, {
  foo: id,
  get children() {
    return $$_clone($$_templ, 1 /* ELEMENT */);
  },
})"
`;

exports[`should compile component with props and spread 1`] = `
"import { $$_create_component, $$_merge_props } from \\"maverick.js/dom\\";
$$_create_component(
  Component,
  $$_merge_props(props, {
    foo: \\"...\\",
    get bar() {
      return id() + 10;
    },
  }),
)"
`;

exports[`should compile component with spread 1`] = `
"import { $$_create_component, $$_merge_props } from \\"maverick.js/dom\\";
$$_create_component(Component, props)"
`;

exports[`should compile component with text children 1`] = `
"import { $$_create_component } from \\"maverick.js/dom\\";
$$_create_component(Component, {
  get children() {
    return \\"foo 10 bar 20 baz\\";
  },
})"
`;

exports[`should compile nested components 1`] = `
"import { $$_create_template, $$_clone, $$_create_component } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><span>Text</span></div>\`),
  $$_templ_2 = /* #__PURE__ */ $$_create_template(\`<div></div>\`);

$$_create_component(Component, {
  get children() {
    return [
      \\"Text\\",
      $$_clone($$_templ, 1 /* ELEMENT */),
      $$_create_component(Foo),
      $$_create_component(Bar, {
        get children() {
          return [$$_clone($$_templ_2, 1 /* ELEMENT */), $$_create_component(Baz)];
        },
      }),
    ];
  },
})
"
`;

exports[`should compile root component 1`] = `
"import { $$_create_component } from \\"maverick.js/dom\\";
$$_create_component(Component)"
`;

exports[`should compile root expression  1`] = `
"import { $$_clone, $$_insert, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);

function Component() {
  return id > 10 ? (() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_insert($$_el, id);

  return $$_el;
})() : 20;
}
  "
`;

exports[`should forward multiple call expressions 1`] = `
"import { $$_clone, $$_insert, $$_create_template, $$_create_component } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`),
  $$_templ_2 = /* #__PURE__ */ $$_templ;

$$_create_component(Component, {
  get children() {
    return [() =>
      (() => {
        const $$_root = $$_clone($$_templ),
          $$_el = $$_root.firstChild;

        $$_insert($$_el, id);

        return $$_el;
      })(), () =>
      (() => {
        const $$_root = $$_clone($$_templ_2),
          $$_el = $$_root.firstChild;

        $$_insert($$_el, id);

        return $$_el;
      })()];
  },
})
"
`;

exports[`should forward single call expression 1`] = `
"import { $$_clone, $$_insert, $$_create_template, $$_create_component } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
$$_create_component(Component, {
  get children() {
    return () =>
      (() => {
        const $$_root = $$_clone($$_templ),
          $$_el = $$_root.firstChild;

        $$_insert($$_el, id);

        return $$_el;
      })();
  },
})"
`;

exports[`should insert fragmented child components 1`] = `
"import { $$_clone, $$_insert, $$_create_component, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><div>Foo</div></div>\`);

(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild,
    $$_el_2 = $$_el.firstChild;

  $$_insert($$_el, $$_create_component(Component), $$_el_2);
  $$_insert($$_el, $$_create_component(Component), $$_el_2);
  $$_insert($$_el, $$_create_component(Component), null);

  return $$_el;
})()"
`;

exports[`should insert multiple child components 1`] = `
"import { $$_clone, $$_insert, $$_create_component, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><div>Foo</div></div>\`);

(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild,
    $$_el_2 = $$_el.firstChild;

  $$_insert($$_el, $$_create_component(Component), $$_el_2);
  $$_insert($$_el, $$_create_component(Component), $$_el_2);
  $$_insert($$_el, $$_create_component(Component), null);

  return $$_el;
})()"
`;
